File: CS/WindowsApplication3/GridHScrollHelper.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Drawing;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace WindowsApplication3 {
    public class GridHScrollHelper {
        GridView view;

        public GridHScrollHelper(GridView view) {
            this.view = view;
        }

        public void EnableScrollByColumns() {
            this.view.LeftCoordChanged += OnLeftCoordChanged;
        }

        void OnLeftCoordChanged(object sender, EventArgs e) {
            if(leftCoordChanged != 0) return;
            BeginScroll();
            try {
                GridColumnInfoArgs firstColumnInfo = GetFirstNonFixedColumnInfo();
                if(firstColumnInfo == null) return;
                int index = ViewInfo.ColumnsInfo.IndexOf(firstColumnInfo);
                if(IsRight) {
                    if(ScrollToNextColumn) {
                        if(index < ViewInfo.ColumnsInfo.Count - 1) {
                            GridColumnInfoArgs nextColumnInfo = ViewInfo.ColumnsInfo[++index];
                            SetLeftCoord(nextColumnInfo.Bounds.Right - Indent - 2, true);
                        }
                    }
                    else
                        SetLeftCoord(firstColumnInfo.Bounds.Right - Indent - 2, true);
                }
                else {
                    if(firstColumnInfo.Column.VisibleIndex != 0)
                        SetLeftCoord(firstColumnInfo.Bounds.Width - firstColumnInfo.Bounds.Right + Indent, false);
                    else
                        SetLeftCoord(view.LeftCoord, false);
                }
                oldLeftCoord = view.LeftCoord;
            }
            finally {
                EndScroll();
            }

        }
        int Indent {
            get { return ViewInfo.ViewRects.FixedLeft.Width + ViewInfo.ViewRects.IndicatorWidth; }
        }

        GridColumnInfoArgs GetFirstNonFixedColumnInfo() {
            GridColumnInfoArgs cArgs = null;
            foreach(GridColumnInfoArgs args in ViewInfo.ColumnsInfo)
                if(args.Column != null && args.Column.Fixed == DevExpress.XtraGrid.Columns.FixedStyle.None
                    && args.Bounds.Left <= Indent)
                    cArgs = args;
            return cArgs;
        }

        private void SetLeftCoord(int delta, bool isRight) {
            if(isRight)
                view.LeftCoord += delta;
            else
                view.LeftCoord -= delta;
        }

        int leftCoordChanged = 0, oldLeftCoord = 0;

        private void EndScroll() {
            leftCoordChanged--;
        }

        private void BeginScroll() {
            leftCoordChanged++;
        }

        GridViewInfo ViewInfo {
            get { return view.GetViewInfo() as GridViewInfo; }
        }

        bool IsRight {
            get { return view.LeftCoord > oldLeftCoord; }
        }

        bool ScrollToNextColumn {
            get {
                GridColumnInfoArgs args = GetFirstNonFixedColumnInfo();
                return args.Bounds.Right <= Indent;
            }
        }

    }
}

File: VB/WindowsApplication3/GridHScrollHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace WindowsApplication3
	Public Class GridHScrollHelper
		Private view As GridView

		Public Sub New(ByVal view As GridView)
			Me.view = view
		End Sub

		Public Sub EnableScrollByColumns()
			AddHandler Me.view.LeftCoordChanged, AddressOf OnLeftCoordChanged
		End Sub

		Private Sub OnLeftCoordChanged(ByVal sender As Object, ByVal e As EventArgs)
			If leftCoordChanged <> 0 Then
				Return
			End If
			BeginScroll()
			Try
				Dim firstColumnInfo As GridColumnInfoArgs = GetFirstNonFixedColumnInfo()
				If firstColumnInfo Is Nothing Then
					Return
				End If
				Dim index As Integer = ViewInfo.ColumnsInfo.IndexOf(firstColumnInfo)
				If IsRight Then
					If ScrollToNextColumn Then
						If index < ViewInfo.ColumnsInfo.Count - 1 Then
							index += 1
							Dim nextColumnInfo As GridColumnInfoArgs = ViewInfo.ColumnsInfo(index)
							SetLeftCoord(nextColumnInfo.Bounds.Right - Indent - 2, True)
						End If
					Else
						SetLeftCoord(firstColumnInfo.Bounds.Right - Indent - 2, True)
					End If
				Else
					If firstColumnInfo.Column.VisibleIndex <> 0 Then
						SetLeftCoord(firstColumnInfo.Bounds.Width - firstColumnInfo.Bounds.Right + Indent, False)
					Else
						SetLeftCoord(view.LeftCoord, False)
					End If
				End If
				oldLeftCoord = view.LeftCoord
			Finally
				EndScroll()
			End Try

		End Sub
		Private ReadOnly Property Indent() As Integer
			Get
				Return ViewInfo.ViewRects.FixedLeft.Width + ViewInfo.ViewRects.IndicatorWidth
			End Get
		End Property

		Private Function GetFirstNonFixedColumnInfo() As GridColumnInfoArgs
			Dim cArgs As GridColumnInfoArgs = Nothing
			For Each args As GridColumnInfoArgs In ViewInfo.ColumnsInfo
				If args.Column IsNot Nothing AndAlso args.Column.Fixed = DevExpress.XtraGrid.Columns.FixedStyle.None AndAlso args.Bounds.Left <= Indent Then
					cArgs = args
				End If
			Next args
			Return cArgs
		End Function

		Private Sub SetLeftCoord(ByVal delta As Integer, ByVal isRight As Boolean)
			If isRight Then
				view.LeftCoord += delta
			Else
				view.LeftCoord -= delta
			End If
		End Sub

		Private leftCoordChanged As Integer = 0, oldLeftCoord As Integer = 0

		Private Sub EndScroll()
			leftCoordChanged -= 1
		End Sub

		Private Sub BeginScroll()
			leftCoordChanged += 1
		End Sub

		Private ReadOnly Property ViewInfo() As GridViewInfo
			Get
				Return TryCast(view.GetViewInfo(), GridViewInfo)
			End Get
		End Property

		Private ReadOnly Property IsRight() As Boolean
			Get
				Return view.LeftCoord > oldLeftCoord
			End Get
		End Property

		Private ReadOnly Property ScrollToNextColumn() As Boolean
			Get
				Dim args As GridColumnInfoArgs = GetFirstNonFixedColumnInfo()
				Return args.Bounds.Right <= Indent
			End Get
		End Property

	End Class
End Namespace
